main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+    *
+    * Invariant: [[SetFormula.ElemSet]], [[SetFormula.Cst]], and [[SetFormula.Var]] must use
+    * disjoint integers (see [[unknowns]]).
+    */
+  sealed trait SetFormula {
+
+    import SetFormula.*
+
+    /** The set of [[Var]] in `this`. */
+    final def variables: SortedSet[Int] = this match {
+      case Univ => SortedSet.empty
+      case Empty => SortedSet.empty
+      case Cst(_) => SortedSet.empty
+      case Var(x) => SortedSet(x)
+      case ElemSet(_) => SortedSet.empty
+      case Compl(t) => t.variables
+      case Inter(_, _, varsPos, _, _, varsNeg, other) =>
+        SortedSet.from(
+          varsPos.iterator.map(_.x) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.variables)
+        )
+      case Union(_, _, varsPos, _, _, varsNeg, other) =>
+        SortedSet.from(
+          varsPos.iterator.map(_.x) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.variables)
+        )
+    }
+
+    /** `true` if `this` contains neither [[Var]] nor [[Cst]]. */
+    final def ground: Boolean = this match {
+      case Univ => true
+      case Empty => true
+      case Cst(_) => false
+      case Var(_) => false
+      case ElemSet(_) => true
+      case Compl(t) => t.ground
+      case Inter(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        cstsPos.isEmpty &&
+          varsPos.isEmpty &&
+          cstsNeg.isEmpty &&
+          varsNeg.isEmpty &&
+          other.forall(_.ground)
+      case Union(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        cstsPos.isEmpty &&
+          varsPos.isEmpty &&
+          cstsNeg.isEmpty &&
+          varsNeg.isEmpty &&
+          other.forall(_.ground)
+    }
+
+    /** Returns all [[Var]] and [[Cst]] that occur in `this`. */
+    final def unknowns: SortedSet[Int] = this match {
+      case Univ => SortedSet.empty
+      case Empty => SortedSet.empty
+      case Cst(c) => SortedSet(c)
+      case ElemSet(_) => SortedSet.empty
+      case Var(x) => SortedSet(x)
+      case Compl(t) => t.unknowns
+      case Inter(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        SortedSet.from(
+          cstsPos.iterator.map(_.c) ++
+            varsPos.iterator.map(_.x) ++
+            cstsNeg.iterator.map(_.c) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.unknowns)
+        )
+      case Union(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        SortedSet.from(
+          cstsPos.iterator.map(_.c) ++
+            varsPos.iterator.map(_.x) ++
+            cstsNeg.iterator.map(_.c) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.unknowns)
+        )
+    }
+
+    /**
+      * Returns the number of connectives in the unary/binary representation of `this`.
+      *
+      * [[Compl]], [[Union]], and [[Inter]] are connectives.
+      */
+    final def size: Int = {
```

Comment:
What's this use for? For the complexity measure in TooComplex etc?

Reply:
yea just debugging and tooComplex